bundles/org.openhab.binding.verisure/src/main/java/org/openhab/binding/verisure/internal/handler/VerisureSmartPlugThingHandler.java

```diff
@@ -0,0 +1,209 @@
+                logger.debug("Trying to set SmartPlug state to {} with URL {} and data {}", operation, url,
+                        queryQLSmartPlugSetState);
+                int httpResultCode = session.sendCommand(url, queryQLSmartPlugSetState, installationId);
+                if (httpResultCode == HttpStatus.OK_200) {
+                    logger.debug("Smartplug state successfully changed!");
+                } else {
+                    logger.warn("Failed to send command, HTTP result code {}", httpResultCode);
+                }
+            }
+        }
+    }
+
+    @Override
+    public synchronized void update(@Nullable VerisureThing thing) {
+        logger.debug("update on thing: {}", thing);
+        updateStatus(ThingStatus.ONLINE);
+        if (getThing().getThingTypeUID().equals(THING_TYPE_SMARTPLUG)) {
+            VerisureSmartPlugs obj = (VerisureSmartPlugs) thing;
+            if (obj != null) {
+                updateSmartPlugState(obj);
+            }
+        } else {
+            logger.warn("Can't handle this thing typeuid: {}", getThing().getThingTypeUID());
+        }
+    }
+
+    private void updateSmartPlugState(VerisureSmartPlugs smartPlugJSON) {
+        Smartplug smartplug = smartPlugJSON.getData().getInstallation().getSmartplugs().get(0);
+        String smartPlugStatus = smartplug.getCurrentState();
+        if (smartPlugStatus != null) {
+            getThing().getChannels().stream().map(Channel::getUID)
+                    .filter(channelUID -> isLinked(channelUID) && !channelUID.getId().equals("timestamp"))
+                    .forEach(channelUID -> {
+                        State state = getValue(channelUID.getId(), smartplug, smartPlugStatus);
+                        updateState(channelUID, state);
+                    });
+            super.update(smartPlugJSON);
+        }
+    }
+
+    public State getValue(String channelId, Smartplug smartplug, String smartPlugStatus) {
+        switch (channelId) {
+            case CHANNEL_SMARTPLUG_STATUS:
+                if ("ON".equals(smartPlugStatus)) {
+                    return OnOffType.ON;
+                } else if ("OFF".equals(smartPlugStatus)) {
+                    return OnOffType.OFF;
+                } else if ("PENDING".equals(smartPlugStatus)) {
+                    // Schedule another refresh.
+                    logger.debug("Issuing another immediate refresh since status is still PENDING ...");
+                    this.scheduleImmediateRefresh(REFRESH_DELAY_SECONDS);
+                }
+                break;
+            case CHANNEL_LOCATION:
+                String location = smartplug.getDevice().getArea();
+                return location != null ? new StringType(location) : UnDefType.NULL;
+            case CHANNEL_HAZARDOUS:
+                return smartplug.isHazardous() ? OnOffType.ON : OnOffType.OFF;
+        }
+        return UnDefType.UNDEF;
+    }
+
+    @NonNullByDefault
+    private static class SmartPlug {
```

Comment:
You have over 25 classes that you are using to represent json objects. I would suggest merging similar classes together as a form of refactoring, possibly putting them in their own standalone files.

But even without refactoring a really good way to simplify and increase readability would be to make the fields for these json classes public and remove any setter methods since they aren't doing any extra logic anyway.

If done correctly data population would make the expection json hierarchy extremely obvious:
```
SmartPlug smartPlugJSON = new SmartPlug();
smartPlugJSON.variables.deviceId = deviceId;
smartPlugJSON.variables.giid = installationId.toString();
smartPlugJSON.variables.state = isOperation;
smartPlugJSON.operationName = "UpdateState";
smartPlugJSON.query = query;
```
But more importantly, removing the setter methods from all the DTOs would remove several hundred lines code 😄

Reply:
I've used a json -> java generator tool :-), but of course I will remove them non-needed setters. Regarding refactoring I like the current approach with one main class that tells which JSON object that is handled within that class.

Reply:
If you remove unneeded setters does that mean that the fields themselves are not needed? Could those be removed as well?